### PR TITLE
fix: PDV 55 add ProvisionedThroughputExceededException handling

### PR DESCRIPTION
**Pull request by BenitoVisone**
This PR aim is to handle the `ProvisionedThroughputExceededException` from `DynamoDB` by wrapping it with a Custom `TooManyRequestException`.

The `ProvisionedThroughputExceededException`  is thrown when the Provisioned Capacity is exceeded and the DynamoDB retry policy reaches the end with no results.

### CODE REVIEW

File: connector-api/src/main/java/it/pagopa/pdv/tokenizer/connector/exception/TooManyRequestException.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package it.pagopa.pdv.tokenizer.connector.exception;
+
+public class TooManyRequestException extends RuntimeException {
+}
```
Comment (rodoviti) [inline connector-api/src/main/java/it/pagopa/pdv/tokenizer/connector/exception/TooManyRequestException.java:4]: refactor rename `TooManyRequestException` into `TooManyRequestsException` and add the exception cause to constructor like the suggested changes
```suggestion
public class TooManyRequestsException extends RuntimeException {

    public TooManyRequestsException(Throwable cause) {
        super(cause);
    }
}
```

**Review by rodoviti — CHANGES_REQUESTED** [review]

**Review by sebbalex — APPROVED** [review]

**Comment by rodoviti** [comment]
@BenitoVisone you can avoid code duplication by implementing "exception mapping" using an aspect (AOP) 😉

File: connector/dao/src/main/java/it/pagopa/pdv/tokenizer/connector/dao/TokenizerConnectorImpl.java [hunk 2]
```diff
@@ -77,6 +79,9 @@ private String saveGlobalToken(String pii) {
                         GlobalFiscalCodeToken.Fields.status);
             }
         }
+        catch(ProvisionedThroughputExceededException e){
+            throw new TooManyRequestsException(e.getCause());
```
Comment (rodoviti) [inline connector/dao/src/main/java/it/pagopa/pdv/tokenizer/connector/dao/TokenizerConnectorImpl.java:83]: you have to set the "current" exception as the cause of the new exception (do this to all refs)
```suggestion
            throw new TooManyRequestsException(e);
```

**Review by rodoviti — CHANGES_REQUESTED** [review]

**Review by rodoviti — APPROVED** [review]
LGTM

PS: if you want to avoid code duplication you can use an aspect (AOP)

**Comment by BenitoVisone** [comment]
> PS: if you want to avoid code duplication you can use an aspect (AOP)

okay I'll try to substitute this behaviour using AOP, thank you 👍

File: connector/dao/src/main/java/it/pagopa/pdv/tokenizer/connector/dao/handler/ConnectorExceptionHandlingAspect.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.aspectj.lang.annotation.AfterThrowing;
+import org.aspectj.lang.annotation.Aspect;
+import org.springframework.stereotype.Component;
+@Slf4j
+@Aspect
+@Component
+public class ConnectorExceptionHandlingAspect {
+
+    @AfterThrowing(pointcut = "execution(* it.pagopa.pdv.tokenizer.connector.dao.*.*(..))", throwing = "ex")
```
Comment (rodoviti) [inline connector/dao/src/main/java/it/pagopa/pdv/tokenizer/connector/dao/handler/ConnectorExceptionHandlingAspect.java:14]: I suggest narrowing the focus of this advice to Connetor implementation(s) only, but i'm not sure the suggest code is correct.....give it a try 😉 
```suggestion
    @AfterThrowing(pointcut = "execution(public it.pagopa.pdv.tokenizer.connector.dao.*(..))", throwing = "ex")
```

**Pull request merged**